Wikipedia/Code/PushNotificationsSettingsViewController.swift

```diff
@@ -0,0 +1,281 @@
+            return UITableViewCell()
+        }
+
+        let item = sections[indexPath.section].items[indexPath.item]
+        cell.title = item.title
+        cell.iconName = item.iconName
+
+        if let themeableCell = cell as Themeable? {
+            themeableCell.apply(theme: theme)
+        }
+
+        if let switchItem = item as? PushNotificationsSettingsSwitchItem {
+            cell.disclosureType = .switch
+            cell.disclosureSwitch.tag = switchItem.tag
+            cell.disclosureSwitch.isOn = switchItem.valueChecker()
+            cell.disclosureSwitch.addTarget(self, action: #selector(userDidTapSwitch(_:)), for: .valueChanged)
+        } else if item is PushNotificationsSettingsActionItem {
+            cell.disclosureType = .externalLink
+        }
+
+        return cell
+    }
+
+    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        guard let header = WMFTableHeaderFooterLabelView.wmf_viewFromClassNib() else {
+            return nil
+        }
+
+        if let themeableHeader = header as Themeable? {
+            themeableHeader.apply(theme: theme)
+        }
+
+        header.text = sections[section].headerText
+        return header
+    }
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let item = sections[indexPath.section].items[indexPath.item] as? PushNotificationsSettingsActionItem else {
+            return
+        }
+
+        item.action()
+        tableView.deselectRow(at: indexPath, animated: true)
+    }
+
+    func tableView(_ tableView: UITableView, shouldHighlightRowAt indexPath: IndexPath) -> Bool {
+        return sections[indexPath.section].items[indexPath.item] as? PushNotificationsSettingsSwitchItem == nil
+    }
+
+    // MARK: - UI Actions
+
+    @objc func userDidTapSwitch(_ sender: UISwitch) {
+        let items = sections.flatMap { section in section.items }.compactMap { item in item as? PushNotificationsSettingsSwitchItem }
+        if let tappedSwitchItem = items.first(where: { item in item.tag == sender.tag }) {
+            tappedSwitchItem.action(sender.isOn)
+        }
+    }
+
+    // MARK: - Themeable
+
+    override func apply(theme: Theme) {
+        super.apply(theme: theme)
+        guard viewIfLoaded != nil else {
+            return
+        }
+        
+        view.backgroundColor = theme.colors.baseBackground
+        tableView.backgroundColor = theme.colors.baseBackground
+        tableView.reloadData()
+    }
+
+}
+
+// MARK: - System push alert/badge permissions and echo subscription/unsubscription
+
+extension PushNotificationsSettingsViewController {
+
+    fileprivate func requestPushPermissions() {
+        notificationsController.requestPermissionsIfNecessary(completionHandler: { (authorized, error) in
+            if authorized {
+                if self.notificationsController.remoteRegistrationDeviceToken == nil {
```

Comment:
So strange! And both the iPhone 7 and iPhone 12 are on iOS 15?

Reply:
I'm still trying to puzzle through the expected OS behavior here – it's frustrating that there isn't a clean synchronous way to get this token. Before I refactor the approach, would you be able to replace `requestPushPermissions()` with the code below and see if it still triggers an error alert on the devices giving you trouble? It may trigger an unexpected switch state, but hopefully not the token missing alert. If you don't have the bandwidth to test this no worries, just lemme know.

```swift
    fileprivate func requestPushPermissions() {
        notificationsController.requestPermissionsIfNecessary(completionHandler: { (authorized, error) in
		
            if authorized {
	        DispatchQueue.main.async {
	             UIApplication.shared.registerForRemoteNotifications()
	        }
            }

            DispatchQueue.main.asyncAfter(deadline: .now() + 5, execute: {
                if authorized {
                    if self.notificationsController.remoteRegistrationDeviceToken == nil {
                        // If we're still awaiting a device token, offer the user the opportunity to retry fetching one
                        DispatchQueue.main.async {
                            let retryAction = UIAlertAction(title: self.echoAlertFailureTryAgainActionTitle, style: .default, handler: { _ in self.requestPushPermissions() })
                            let cancelAction = UIAlertAction(title: CommonStrings.cancelActionTitle, style: .cancel, handler: { _ in self.updateSections() })
                            self.wmf_showAlert(title: self.echoAlertFailureTitle, message: self.echoAlertFailureMessage, actions: [retryAction, cancelAction], completion: {
                                // Silently trigger a remote registration request to fetch a device token
                                UIApplication.shared.registerForRemoteNotifications()
                            })
                        }
                    } else {
                        // User is authorized for on device push alerts and is now awaiting subscription to Echo Notifications
                        self.subscribeToEchoNotifications()
                    }
                } else {
                    // User isn't authorized, just update sections to a failure state
                    self.updateSections()
                }
            })

        })
    }

```

Reply:
Sure, I just tested this - I didn't see the alert on the new devices:

![iPhone12_2](https://user-images.githubusercontent.com/3620196/156789559-f8377abb-fbc8-4ed9-b02c-41e3e4241d1e.gif)

Just FYI, I was curious as to why the toggle switches back to off during the 5 second delay - turns out accepting the permissions modal fires the `didBecomeActiveNotification` notification, which calls `updateSections` again. I also tried bumping this delay down - I was able to get to 0.4 seconds without seeing the error alert, though I did still see the toggle briefly switch to Off due to the `didBecomeActive` notification.

Reply:
> Just FYI, I was curious as to why the toggle switches back to off during the 5 second delay - turns out accepting the permissions modal fires the `didBecomeActiveNotification` notification

Ah, I hadn't noticed the very specific language from the API docs re: `didBecomeActiveNotification`: `An app is active when it is receiving events. It ... loses focus when an overlay window pops up.` I think `willEnterForegroundNotification` might fit our purposes here, what do you think?

Reply:
@staykids sure, I think it's worth a try. I always figured these notifications acted the same, just fired at slightly different moments in time. But the `willEnterForegroundNotification` documentation doesn't seem to mention overlay windows, so maybe that's a difference too. Let me know if you need me to test it on one of my devices.

Reply:
I pushed a few new commits and I think this is ready for another look.

- Immediately call `UIApplication.shared.registerForRemoteNotifications()` when system permissions are authorized
- Fixed a memory leak in `updateSections()`
- Switched to use the foreground notification instead of did become active

I'm trying every approach I can before refactoring the permissions/token/subscription flow because it's pretty clean now 😅, but if you're still running into issues with this lemme know and I'll start the process of breaking it apart. Apologies for the back and forth on this PR – I wish I had a problem device to repro the issues you're running into to save you some testing.

I also noticed I'll probably need to do a bit more work with my log out unsubscription approach, but it'll require a little research so I may spin up a new Phab to address it separately.

Reply:
@tonisevener Alrighty got some updated work for you to take a look at. I didn't quite execute this the same way we discussed (I still think the approach you suggested is much cleaner from an API standpoint than my approach here), but was running into some irritating containment constraints (because of `WMFNotificationsController`'s inclusion in the WMF framework) so just decided to generalize out some of the retry work and apply it directly in `PushNotificationsSettingsViewController`. I do not enjoy all of my nesting and dispatch spaghetti, but so far it seems pretty functional and directly addresses the problem of potentially having to wait a bit for a device token. 

A summary:
- Adds a generalized `RetryBlockTask` that executes a block up to a set number of times, exiting early if that block succeeds
- Removes `WMFNotificationsController` token delegate code
- Uses `RetryBlockTask` as part of the device token flow, continuing to echo subscription if a device token is available after up to 3 retry checks for it, or presenting an alert allowing the user to retry the process.

🤞🏽 Hopefully this gracefully resolves the device token related issues you were running into, but of course if not I'll jump back into it.

Reply:
@tonisevener Actually - doing another test pass of it I'm still noticing some awkward behavior. Let me do another pass before you take a look.

Reply:
@tonisevener I updated this to also block interaction with the primary push switch if an echo subscription/unsubscription is in progress. Ready for a look!